packages/@uppy/aws-s3/src/MiniXHRUpload.js

```diff
@@ -0,0 +1,360 @@
+            uploader: this,
+            bytesUploaded: ev.loaded,
+            bytesTotal: ev.total
+          })
+        }
+      })
+
+      xhr.addEventListener('load', (ev) => {
+        this.uppy.log(`[AwsS3/XHRUpload] ${id} finished`)
+        timer.done()
+        queuedRequest.done()
+        if (this.uploaderEvents[file.id]) {
+          this.uploaderEvents[file.id].remove()
+          this.uploaderEvents[file.id] = null
+        }
+
+        if (opts.validateStatus(ev.target.status, xhr.responseText, xhr)) {
+          const body = opts.getResponseData(xhr.responseText, xhr)
+          const uploadURL = body[opts.responseUrlFieldName]
+
+          const uploadResp = {
+            status: ev.target.status,
+            body,
+            uploadURL
+          }
+
+          this.uppy.emit('upload-success', file, uploadResp)
+
+          if (uploadURL) {
+            this.uppy.log(`Download ${file.name} from ${uploadURL}`)
+          }
+
+          return resolve(file)
+        } else {
+          const body = opts.getResponseData(xhr.responseText, xhr)
+          const error = buildResponseError(xhr, opts.getResponseError(xhr.responseText, xhr))
+
+          const response = {
+            status: ev.target.status,
+            body
+          }
+
+          this.uppy.emit('upload-error', file, error, response)
+          return reject(error)
+        }
+      })
+
+      xhr.addEventListener('error', (ev) => {
+        this.uppy.log(`[AwsS3/XHRUpload] ${id} errored`)
+        timer.done()
+        queuedRequest.done()
+        if (this.uploaderEvents[file.id]) {
+          this.uploaderEvents[file.id].remove()
+          this.uploaderEvents[file.id] = null
+        }
+
+        const error = buildResponseError(xhr, opts.getResponseError(xhr.responseText, xhr))
+        this.uppy.emit('upload-error', file, error)
+        return reject(error)
+      })
+
+      xhr.open(opts.method.toUpperCase(), opts.endpoint, true)
+      // IE10 does not allow setting `withCredentials` and `responseType`
+      // before `open()` is called.
+      xhr.withCredentials = opts.withCredentials
+      if (opts.responseType !== '') {
+        xhr.responseType = opts.responseType
+      }
+
+      Object.keys(opts.headers).forEach((header) => {
+        xhr.setRequestHeader(header, opts.headers[header])
+      })
+
+      const queuedRequest = this.requests.run(() => {
+        xhr.send(data)
+        return () => {
+          timer.done()
+          xhr.abort()
+        }
+      }, { priority: 1 })
+
+      this._onFileRemoved(file.id, () => {
+        queuedRequest.abort()
+        reject(new Error('File removed'))
+      })
+
+      this._onCancelAll(file.id, () => {
+        queuedRequest.abort()
+        reject(new Error('Upload cancelled'))
+      })
+    })
+  }
+
+  _uploadRemoteFile (file, current, total) {
+    const opts = this._getOptions(file)
+    return new Promise((resolve, reject) => {
+      this.uppy.emit('upload-started', file)
```

Comment:
should we leave this event emit to `index.js` as done [here](https://github.com/transloadit/uppy/pull/2060/files#diff-57355ea8780221946e05753c979ce54eR138)